core/templates/pages/exploration-editor-page/services/exploration-warnings.service.ts

```diff
@@ -488,6 +530,57 @@ export class ExplorationWarningsService {
   updateWarnings(): void {
     this._updateWarningsList();
   }
+
+  redirectionIsValid(
+      sourceStateName: string, destStateName: string, links: GraphLink[]
+  ): boolean {
+    let distance = this.getDistanceToDestState(
+      sourceStateName, destStateName, links);
+    // Raise validation error if the creator redirects the learner
+    // back by more than MAX_CARD_COUNT_FOR_VALID_REDIRECTION cards.
+    if (distance > AppConstants.MAX_CARD_COUNT_FOR_VALID_REDIRECTION) {
+      return false;
+    } else {
+      return true;
+    }
+  }
+
+  getDistanceToDestState(
+      sourceStateName: string, destStateName: string, links: GraphLink[]
+  ): number {
+    let distance = -1;
+    let stateFound = false;
+    let queue: string[] = [];
+    let seen: Record<string, boolean> = {};
+    seen[sourceStateName] = true;
+    queue.push(sourceStateName);
+    while (queue.length > 0 && !stateFound) {
+      let queueSize = queue.length;
+      distance++;
+      while (queueSize > 0) {
+        // '.shift()' here can return an undefined value, but we're already
+        // checking for queue.length > 0, so this is safe.
+        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+        let currStateName = queue.shift()!;
```

Comment:
Shift on array is O(n), please use deque for constant time pop.

Reply:
Makes sense. Will do.